### ISSUE TITLE
Need information about Player.isPlaying()

### ISSUE THREAD

**Issue by leemoxx** [google/ExoPlayer#6766]
### [REQUIRED] Question
I am trying to understand why Player.isPlaying() was introduced in v2.10.5. I mean getPlayWhenReady() served the same purpose on the interface. From what I remember (I could be wrong), the method was called to keep the current time on the player widget up to date. Comparing both .5 and .4, I have noticed that isPlaying() is now called to perform that (I think).

The reason I am concerned about this is that I use the ExoPlayer widget with multiple players (ExoPlayer, MediaPlayer, etc.) So, when I have switched to .5, the widget did not work properly. 

On .4 I have made checks on `public boolean getPlayWhenReady()` and if ExoPlayer was selected, I would return `mExoPlayer.getPlayWhenReady()` while handling the rest of the players playing state with other conditions (mMediaPlayer.isPlaying() etc.). 

On .5 I have kept the same checks and on `public boolean getPlayWhenReady()`, I have added `mExoPlayer.isPlaying()` to `public boolean isPlaying()` but seeking resulted in the seekbar going to the start position then moving to the new location. That is what confused me.

I have decided to use `mExoPlayer.getPlayWhenReady()` (inside `public boolean isPlaying()`) instead `mExoPlayer.isPlaying()` and the seekbar worked perfectly. Now, the code inside both `public boolean isPlaying()` and `public boolean getPlayWhenReady()` is the same.

Is this the right approach for supporting the seekbar positioning for multiple players i.e. placing the same code block in both `public boolean getPlayWhenReady()` and `public boolean isPlaying()`?

### [REQUIRED] Searched documentation and issues
I have referred to the javadocs:

https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/Player.html#isPlaying--

Though, it does not offer any explanation or comparison between `isPlaying()` and `getPlayWhenReady()` My app only plays offline files (no streaming or anything).

### Link to test content

**Comment by Samrobbo** [comment]
Hi, #6203 was the catalyst for introducing these changes, and the discussion in the comments covers the key reasons for `Player.isPlaying()`. @tonihei was responsible for the change, so I'll let him answer your question regarding whether that is the right approach.

**Comment by tonihei** [comment]
It's slightly unclear to me what you are trying to do and where the problem is exactly. Could you specify what the issue is more clearly in terms of what (a) I need to do to reproduce the problem, (b) what you expect to happen, (c) what actually happens?

Besides that please note that `isPlaying` is a super-set of `playWhenReady` as documented in the javadoc link you already provided yourself. So it's not a replacement and it's not meant to be one.

**Comment by leemoxx** [comment]
@Samrobbo Hello and thanks for the help.
@tonihei Hi. I am using the `ExoPlayer` playback widget to support multiple players. I do this by creating my own `Player` implementation and forwarding calls from this implementation to my instance of `SimpleExoPlayer` (when ExoPlayer is selected). 

For example, inside `public boolean getPlayWhenReady()` of  my`Player` implementation, I do `mExoPlayer.getPlayWhenReady()` where `mExoPlayer` is my `SimpleExoPlayer` instance. Then, I handle the rest of the players checks in the same method. I repeat for all methods of the interface. This worked without issues until 2.10.4

Starting 2.10.5, where `isPlaying()` was introduced, I tried adding `return mExoPlayer.isPlaying()` within `public boolean isPlaying()` but that caused the seekbar to flicker because it seeked (the seekbar only) to the beginning of the file, then back to the destination time selected. When I switched to `return mExoPlayer.getPlayWhenReady()` inside `public boolean isPlaying()` everything worked perfectly.

What I am trying understand is why the widget behaves like that when I return `return mExoPlayer.isPlaying()` Also, is my approach of using `return mExoPlayer.getPlayWhenReady()` within `isPlaying()` the right approach?

**Comment by tonihei** [comment]
> Also, is my approach of using return mExoPlayer.getPlayWhenReady() within isPlaying() the right approach?

That's not definitely not correct and you should be able to forward `mExoPlayer.isPlaying()`. 

> that caused the seekbar to flicker because it seeked (the seekbar only) to the beginning of the file, then back to the destination time selected. 

Which seekbar are you talking about exactly?  None of our UI components is using `isPlaying()` in a way that should influence the seekbar position.

**Comment by leemoxx** [comment]
Thanks for taking the time to help me, tonihei. I am using `PlayerControlView`

**Comment by tonihei** [comment]
I can't reproduce that with in our demo app with a random example stream and doing some seeking. Note that `SimpleExoPlayer` is also forwarding the calls to another `Player` implementation, so it's conceptually the same as in your setup. 

Could you please try to provide complete reproduction steps including example media  that would allow us to reproduce the problem? Otherwise it's not possible to understand whether this is a bug in our library, a media issue, or a bug in your implementation code.

**Comment by leemoxx** [comment]
I am currently testing this on a Lollipop device. I will see if this happens on Android Q (currently acquiring the emulator). My app is an **audio player**. I have used a lot of tracks with the same issue. I will report back with further information after testing my app on Android Q.

Regards and thanks.

**Comment by leemoxx** [comment]
The same thing happens on Android Q. This is a bit frustrating. Does the ExoPlayer demo use `PlayerControlView` Also, does it load audio files?

My `Player` implementation currently (after simplification) has the following methods implementation:

```
            @Override
            public boolean isPlaying() {
            	return mExoPlayer.isPlaying(); // If I switch this to mExoPlayer.getPlayWhenReady() everything works perfectly.
            }

            @Override
            public boolean getPlayWhenReady() {
            	return mExoPlayer.getPlayWhenReady();
            }
```

**Comment by tonihei** [comment]
ExoPlayer's demo app uses both. Could you just try to play your file in our demo app or send us your file so that we can check where the problem is coming from?

**Comment by leemoxx** [comment]
OK, I have examined my code once again. After a lot of debugging, I believe I have found what caused this. I had a check on `public long getContentPosition() {` I have checked if the player was playing and I have returned `mPlayer.getContentPosition();` when it was playing, and 0 if it did not. I have done this to combat an issue from one of the previous releases.

I will make sure that everything is working properly and I will close the issue. Thank you so much for the help, @tonihei
